File: src/limitless_cli/output/formatters.py
"""Output formatting utilities migrated from the monolithic script."""

from __future__ import annotations

import json
import sys
from typing import Iterator, Dict, Any

__all__: list[str] = ["stream_json", "print_markdown"]


def stream_json(gen: Iterator[Dict[str, Any]]) -> None:
    """Write an iterator of JSON-able dicts as a compact JSON array."""

    sys.stdout.write("[")
    first = True
    for item in gen:
        if not first:
            sys.stdout.write(",")
        json.dump(item, sys.stdout, separators=(",", ":"))
        first = False
    sys.stdout.write("]\n")
    sys.stdout.flush()


def print_markdown(gen: Iterator[Dict[str, Any]]) -> None:
    """Extract and print the markdown field of each lifelog."""

    for item in gen:
        md = item.get("markdown") or item.get("data", {}).get("markdown", "")
        if md:
            print(md) 

File: internal/output/formatters.go
// Package output provides output formatting utilities for the Limitless CLI.
package output

import (
	"encoding/json"
	"fmt"
	"os"
)

// StreamJSON writes an iterator of JSON-able maps as a compact JSON array.
func StreamJSON(logs <-chan map[string]interface{}) {
	fmt.Print("[")
	first := true
	for item := range logs {
		if !first {
			fmt.Print(",")
		}
		data, err := json.Marshal(item)
		if err != nil {
			continue
		}
		os.Stdout.Write(data)
		first = false
	}
	fmt.Println("]")
}

// StreamJSONSlice writes a slice of maps as a compact JSON array.
func StreamJSONSlice(logs []map[string]interface{}) {
	fmt.Print("[")
	for i, item := range logs {
		if i > 0 {
			fmt.Print(",")
		}
		data, err := json.Marshal(item)
		if err != nil {
			continue
		}
		os.Stdout.Write(data)
	}
	fmt.Println("]")
}

// PrintMarkdown extracts and prints the markdown field of each lifelog.
func PrintMarkdown(logs <-chan map[string]interface{}) {
	for item := range logs {
		md := ""
		if m, ok := item["markdown"].(string); ok && m != "" {
			md = m
		} else if data, ok := item["data"].(map[string]interface{}); ok {
			if m, ok := data["markdown"].(string); ok {
				md = m
			}
		}
		if md != "" {
			fmt.Println(md)
		}
	}
}

// PrintMarkdownSlice extracts and prints the markdown field from a slice.
func PrintMarkdownSlice(logs []map[string]interface{}) {
	for _, item := range logs {
		md := ""
		if m, ok := item["markdown"].(string); ok && m != "" {
			md = m
		} else if data, ok := item["data"].(map[string]interface{}); ok {
			if m, ok := data["markdown"].(string); ok {
				md = m
			}
		}
		if md != "" {
			fmt.Println(md)
		}
	}
}

// PrintJSON prints a single item as formatted JSON.
func PrintJSON(item interface{}) {
	data, err := json.MarshalIndent(item, "", "  ")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error encoding JSON: %v\n", err)
		return
	}
	fmt.Println(string(data))
}

